### PR TITLE
Drop old code based on python < 3.7, part 1: eventlet directory

### PR DESCRIPTION
**Pull request by eli-schwartz**
@jayofdoom successfully nerdsniped me, so I nerdsniped him back by teaching him about pyupgrade and ruff for automated migration. Here's a nice chunk of changes, which has the advantage of dropping a dependency on the "six" module.

Automatically upgrade code to drop python2 specific logic. Done in three passes:
- automatically run `pyupgrade --py3-only --keep-percent-format` on all files, to update to python 3.0 without changing percent formatting to `.format()` (this would be fairly noisy and isn't even 100% desired by everyone)
- manually fix up a bunch of six specific stuff, mainly dropping unused "import six" left behind by pyupgrade, and drop six from install requires.
- switch over to `ruff check` with its `--select UP*` class of rules targeting py37, and apply autofixes.

### CODE REVIEW

**Comment by jstasiak** [comment]
If this is at least partially a result of some automated migration – can you share the exact commands you ran (and any other relevant context)?

It'll help in reviewing this and if/when someone needs to do something like this in the future (knowledge sharing).

I'd put it in the commit message when merging.

**Comment by eli-schwartz** [comment]
> If this is at least partially a result of some automated migration – can you share the exact commands you ran (and any other relevant context)?



> I'd put it in the commit message when merging.

It's in the commit message :) that's one of my rules when I do this sort of thing.

**Comment by jstasiak** [comment]
Oops, sorry – very nice (although I'm not sure about Markdown in there, outside of GH it's noise).

Let me swap "commit message" for "commit message and the PR description" then :)

**Comment by eli-schwartz** [comment]
And to also share a bit of knowledge I told Jay in IRC...

The main advantage of pyupgrade over ruff check --fix is that ruff didn't implement fixers for migrating off of six (and pyupgrade's aren't complete). Other than that, and particularly for gating on versions starting with python 3.7, both tools are more or less equal in value in my experience other than ruff being *much* less opinionated because ruff allows you to pick and choose which fixer rules you want. Pyupgrade is kind of a pain for that (and also has a weird insistence on passing each filename by hand, because the author thinks that you should only be allowed to run tools via his "pre-commit.ci" tool).

**Comment by 4383** [comment]
Nice changes thanks.
I still have to finish to review them.

File: eventlet/debug.py
Comment (4383) [inline eventlet/debug.py:15]: raw string looks correct

File: eventlet/green/http/__init__.py
Comment (4383) [inline eventlet/green/http/__init__.py:56]: `assert` must not be used in code loaded at runtime, if user call python with the optimization flag, all asserts would be removed, and, so, this kind of code would be removed at runtime.

That's one less to transform.

File: eventlet/green/http/client.py
Comment (4383) [inline eventlet/green/http/client.py:1129]: The current version LGTM, but, since we support versions of python higher than 3.7,  f-string could be used for this kind code.

File: eventlet/green/ssl.py
Comment (4383) [inline eventlet/green/ssl.py:128]: This section (including the following lines), should be removed.

File: eventlet/green/ssl.py
Comment (4383) [inline eventlet/green/ssl.py:133]: That's specific to Python versions lower than 3.7, let's drop that code.

**Review by 4383 — CHANGES_REQUESTED** [review]
I still have files to reviews, but some part of the current code should be updated, see my inline comments.

File: eventlet/green/subprocess.py
Comment (4383) [inline eventlet/green/subprocess.py:111]: Not sure to see why these calls were duplicated... doesn't matter now that they are removed

File: eventlet/websocket.py
Comment (4383) [inline eventlet/websocket.py:429]: I would suggest taking advantage of the situation to replace this “assertion” with a condition. This is a module loaded at runtime, so if optimization is used this code will be removed.

Else it should be rework with a follow up patch.

File: eventlet/zipkin/http.py
Comment (4383) [inline eventlet/zipkin/http.py:39]: I wonder if it would be more convenient to remove this entire module. I'm not sure, but the warning message below seems to suggest that zipkin is not supported by Python 3+.

Thoughts?

File: tests/green_http_test.py
Comment (4383) [inline tests/green_http_test.py:4]: I think we could simply remove that line. Pytest would ignore test if `__test__ = False`, so I don't think this line is necessary.

It reflect a time where tests would had to be ignored with python 2.7

File: tests/isolated/patcher_open_kwargs.py
Comment (4383) [inline tests/isolated/patcher_open_kwargs.py:7]: text mode is the default, LGTM

**Comment by 4383** [comment]
I now reviewed all the files, I've no other comments than those I already made, please have a look.
Thanks

**Comment by 4383** [comment]
Also, considering the heavy size of this patch, I'd love another double check from maintainers, I may missed some issues in my review.

**Comment by itamarst** [comment]
I am pretty worried about this PR so please do NOT merge it until I have time (later today) to validate my concerns and write them up here.

File: eventlet/websocket.py
Comment (eli-schwartz) [reply eventlet/websocket.py:429]: From a hasty look at the context, what you're saying makes sense.

Asserts are for cases where the matched condition indicates your own code is incorrectly written. They "cannot happen" because as everyone knows, we ourselves write bug-free code, so why check it.

This code  *appears* to be parsing external inputs, which means other people's code is buggy and we have to check that :D or well, innumerable things including fuzzer attacks.

I can fix it with a follow-up patch.

File: eventlet/zipkin/http.py
Comment (eli-schwartz) [reply eventlet/zipkin/http.py:39]: I wondered about that but lacked confidence to delete it outright since it might for example have a TODo intention to port it.

Comment (4383) [reply eventlet/zipkin/http.py:39]: I've no idea if someone else wanted to port it.
I'd personally suggest to remove it, but, maybe we want some deprecation period first, though, if I'm right, it won't work with python 3...

File: eventlet/websocket.py
Comment (4383) [reply eventlet/websocket.py:429]: thanks

**Comment by itamarst** [comment]
Here are my concerns, after reviewing the PR a bit more I am not sure that they are too big of an issue, but still worth considering.

### Problem: Eventlet is special

Eventlet is a weird library that _emulates_ the Python standard library, currently for 2.7 through 3.12. The auto-upgrade tools don't know about this scenario at all, they're assuming you're just a user of APIs.

Consider this code snippet, that was part of @jayofdoom's motivation, since it apparently caused breakage for someone: https://github.com/eventlet/eventlet/blob/4da815f51ba7afb1b476b8c0f539da1c8ba3e15a/eventlet/green/ssl.py#L140-L148

The auto-upgrade stuff doesn't touch it. There are likely other unconditional pieces of code only there to emulate Python 2 stdlib, or Python 3.5 stdlib, etc..

And by merging this PR, we take away the context that lets us figure out _why_ that code is doing what it does. In this case there's a comment saying it's for Python 2, but my worry was that there wouldn't always be.

In practice, I skimmed the PR and didn't see many other cases like this, but it's very long and I may have missed those...

### Problem: Code and tests in a single pass

Given how dynamic Python is, we need to rely heavily on tests for API stability. However, this PR modifies both code and tests in one go.

A better procedure would be to upgrade all the code, run the tests and make sure they pass, and _then_ upgrade the tests.

**Comment by 4383** [comment]
> Given how dynamic Python is, we need to rely heavily on tests for API stability. However, this PR modifies both code and tests in one go.
> 
> A better procedure would be to upgrade all the code, run the tests and make sure they pass, and _then_ upgrade the tests.

+1, also, that would reduce the number of files to review, and, so decrease the chances to miss something. Could be done by  simply unstagging changes temporarly from git and by committing both paths (`eventlet/` and `tests/`) independently.

**Comment by eli-schwartz** [comment]
I agree it makes sense to split up the affected changes into two commits, one just for the tests. Small caveat: GitHub Actions won't test them, it only runs on the branch head. :P Would you prefer it in two PRs? Something else?

**Comment by jayofdoom** [comment]
Two PRs is the easiest way to demonstrate it, for sure.

**Comment by itamarst** [comment]
> I agree it makes sense to split up the affected changes into two commits, one just for the tests. Small caveat: GitHub Actions won't test them, it only runs on the branch head. :P Would you prefer it in two PRs? Something else?

The tests seem to run on PRs?

**Comment by eli-schwartz** [comment]
> The tests seem to run on PRs?

It does not bisect a PR and check that tests pass at each step of the way. This isn't Gerrit, PRs cannot be treated as a series of independent patches in-series.

**Comment by itamarst** [comment]
Oh, I see. Yeah, two PRs is the way to go.

**Comment by eli-schwartz** [comment]
Okay, I separated them out and also spent some time breaking it up even more. Now I ran ruff per-fixer upgrades one at a time, and  used pyupgrade at the end -- mostly six-based stuff.

**Review by 4383 — APPROVED** [review]
Latest patch set LGTM, waiting for @itamarst's feedback.

**Comment by 4383** [comment]
A couple of PEP8 issue need to be addressed.

**Review by itamarst — APPROVED** [review]

**Comment by itamarst** [comment]
Just the style changes left, yeah.

**Comment by eli-schwartz** [comment]
Style changes fixed up and absorbed into the relevant commits (which have been updated to note that they aren't a straight commit of the results anymore).

**Review by 4383 — APPROVED** [review]
Latest patch set LGTM

**Comment by eli-schwartz** [comment]
Thanks for merging.

Aside: in general I'd appreciate knowing in advance if a PR is going to be merged via squashing.

**Comment by itamarst** [comment]
> Thanks for merging.
> 
> Aside: in general I'd appreciate knowing in advance if a PR is going to be merged via squashing.

Thank you for contributing, and apologies if lack of that knowledge caused you extra work.

**Comment by eli-schwartz** [comment]
It's not about the extra work as much as it is, that if I'm aware in advance then there are some things I can do to partially mitigate the extremity level of the resulting damage.

Once the squash merge has happened, it is too late -- the damage is done, there is no going back.

**Pull request merged**
